llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -961,23 +961,27 @@ getSizeFromDwarfHeaderAndSkip(BinaryStreamReader &Reader) {
 
 /// Given a list of MCFragments, return a vector with the concatenation of their
 /// data contents.
-/// If any fragment is not an MCDataFragment, an error is returned.
+/// If any fragment is not an MCDataFragment, or the fragment is an
+/// MCDwarfLineAddrFragment and the section containing that fragment is not a
+/// debug_line section, an error is returned.
 static Expected<SmallVector<char, 0>>
-mergeDataFragmentContents(const MCSection::FragmentListType &FragmentList) {
+mergeMCFragmentContents(const MCSection::FragmentListType &FragmentList,
+                        bool IsDebugLineSection = false) {
   SmallVector<char, 0> mergedData;
-  for (const MCFragment &Fragment : FragmentList) {
-    if (const auto *DataFragment = dyn_cast<MCDataFragment>(&Fragment))
-      llvm::append_range(mergedData, DataFragment->getContents());
+  for (const MCFragment &F : FragmentList) {
+    if (F.getKind() == llvm::MCFragment::FT_Data)
+      llvm::append_range(mergedData, cast<MCDataFragment>(F).getContents());
```

Comment:
Why did we change this from a dyn_cast to a "Kind" test followed by a cast? The dyn_cast approach seems more idiomatic to what the LLVM codebase uses. 

I could be missing some details here, but this change seems unrelated to what the PR is trying to accomplish.

Reply:
Two reasons:

1. Look at the if statement below, I need to also check to make sure there is an `MCDwarfLineAddrFragment` only if the section is a debug_line section. 

2. Imo, a check for the `Kind` is more efficient than the cast

Reply:
> Imo, a check for the Kind is more efficient than the cast

They are virtually the same thing for builds with assertions. So the current form is doing the same check twice (a `cast` will check too).

> need to also check to make sure there is an MCDwarfLineAddrFragment

Wouldn't this do the job?

```cpp
else if (const auto *LineFragment = dyn_cast<MCDwarfLineAddrFragment>(&Fragment))
  if(IsDebugLineSection)
    llvm::append_range(...)
```

Reply:
It would, but when I wrote I just went a different way, especially because if you see any of the code in `MCAssembler.cpp` anything to do with the fragment switches on the fragment kind so I want to keep that consistent. If it could all be achieved with a cast, then why add a Kind to the class anyway? If the dyn_cast is more idiomatic to the llvm coding style, then I do not know why MCAssembler uses the Kind in a switch.

Reply:
I changed it to a dyn_cast :)

Reply:
Awesome, thank you!

> why MCAssembler uses the Kind in a switch.

Because they need a switch ;)
C++ doesn't allow switching on a type, more recently LLVM gained the ability to do so with some template tricks. See for example https://llvm.org/doxygen/TypeSwitch_8h_source.html#l00107